Source/WebCore/Modules/url-pattern/URLPatternParser.cpp

```diff
@@ -0,0 +1,341 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternParser.h"
+
+#include "URLPatternTokenizer.h"
+
+namespace WebCore {
+namespace URLPatternUtilities {
+
+URLPatternParser::URLPatternParser(EncodingCallbackType urlComponentType, String&& segmentWildcardRegexp)
+    : m_callbackType(urlComponentType)
+    , m_segmentWildcardRegexp(segmentWildcardRegexp)
+{
+}
+
+ExceptionOr<void> URLPatternParser::performParse(const URLPatternStringOptions& patternSettings)
+{
+    ExceptionOr<void> maybeFunctionException;
+
+    while (m_index < m_tokenList.size()) {
+        auto charToken = tryToConsumeToken(TokenType::Char);
+        auto nameToken = tryToConsumeToken(TokenType::Name);
+        auto regexOrWildcardToken =  tryToConsumeRegexOrWildcardToken(nameToken);
+
+        if (nameToken.isInitialized() || regexOrWildcardToken.isInitialized()) {
+            String prefix = emptyString();
+
+            if (charToken.isInitialized())
+                prefix = charToken.value.toString();
+
+            if (!prefix.isEmpty() && prefix != patternSettings.prefixCodepoint) {
+                m_pendingFixedValue = makeString(WTFMove(m_pendingFixedValue), prefix);
+                prefix = emptyString();
+            }
+
+            maybeFunctionException = maybeAddPartFromPendingFixedValue();
+            if (maybeFunctionException.hasException())
+                return maybeFunctionException.releaseException();
+
+            auto modifierToken = tryToConsumeModifierToken();
+
+            maybeFunctionException = addPart(WTFMove(prefix), nameToken, regexOrWildcardToken, ""_s, modifierToken);
+            if (maybeFunctionException.hasException())
+                return maybeFunctionException.releaseException();
+
+            continue;
+        }
+
+        auto fixedToken = charToken;
+
+        if (fixedToken.isNull())
+            fixedToken = tryToConsumeToken(TokenType::EscapedChar);
+        else if (fixedToken.isInitialized()) {
+            m_pendingFixedValue = makeString(WTFMove(m_pendingFixedValue), fixedToken.value);
```

Comment:
I recommend using std::exchange instead of `WTFMove()` here since `m_pendingFixedValue` is likely getting re-used after the move (which is not a great practice).